text/0000-dst-coercion.md

```diff
@@ -0,0 +1,182 @@
+```
+
+Note that there are some coercions which are not given by `CoerceUnsized`, e.g.,
+from safe to unsafe function pointers, so it really is a `CoerceUnsized` trait,
+not a general `Coerce` trait.
+
+
+## Compiler checking
+
+### On encountering an implementation of `CoerceUnsized` (type collection phase)
+
+* If the impl is for a built-in pointer type, we check nothing, otherwise...
+* The compiler checks that the `Self` type is a struct or tuple struct and that
+the `Target` type is a simple substitution of type parameters from the `Self`
+type (one day, with HKT, this could be a regular part of type checking, for now
+it must be an ad hoc check). We might enforce that this substitution is of the
+form `X/Y` where `X` and `Y` are both formal type parameters of the
+implementation (I don't think this is necessary, but it makes checking coercions
+easier and is satisfied for all smart pointers).
+* The compiler checks each field in the `Self` type against the corresponding field
+in the `Target` type. Assuming `Fs` is the type of a field in `Self` and `Ft` is
+the type of the corresponding field in `Target`, then either `Ft <: Fs` or
+`Fs: CoerceUnsized<Ft>` (note that this includes built-in coercions).
+* There must be only one field that is coerced.
+* We record for each impl, the index of the field in the `Self` type which is
+coerced.
+
+### On encountering a potential coercion (type checking phase)
+
+* If we have an expression with type `E` where the type `F` is required during
+type checking and `E` is not a subtype of `F`, nor is it coercible using the
+built-in coercions, then we search for a bound of `E: CoerceUnsized<F>`. Note
+that we may not at this stage find the actual impl, but finding the bound is
+good enough for type checking.
+
+* If we require a coercion in the receiver of a method call or field lookup, we
+perform the same search that we currently do, except that where we currently
+check for coercions, we check for built-in coercions and then for `CoerceUnsized`
+bounds. We must also check for `Unsize` bounds for the case where the receiver
+is auto-deref'ed, but not autoref'ed.
+
+
+### On encountering an adjustment (translation phase)
+
+* In trans (which is post-monomorphisation) we should always be able to find an
+impl for any `CoerceUnsized` bound. 
+* If the impl is for a built-in pointer type, then we use the current coercion
+code for the various pointer kinds (`Box<T>` has different behaviour than `&` and
+`*` pointers).
+* Otherwise, we lookup which field is coerced due to the opt-in coercion, move
+the object being coerced and coerce the field in question by recursing (the
+built-in pointers are the base cases).
+
+
+### Adjustment types
+
+We add `AdjustCustom` to the `AutoAdjustment` enum as a placeholder for coercions
+due to a `CoerceUnsized` bound. I don't think we need the `UnsizeKind` enum at
+all now, since all checking is postponed until trans or relies on traits and impls.
```

Comment:
Yes! I've done this in a local branch and now there's only unsizing coercion creation (typeck) and unsizing information generation (trans) that cares about the specific kind of unsizing being performed.
The later is even `O(1)`, it ignores structure nesting completely.
This is how it looks like (I hope to replace `root_target` by the resulting pointer type):
```rust
// In the case of `&Struct<[T; N]>` -> `&Struct<[T]>`, the types are:
// * leaf_source: `[T; N]`
// * leaf_target: `[T]`
// * root_target: `Struct<[T]>`
pub struct AutoUnsize<'tcx> {
    leaf_source: Ty<'tcx>,
    leaf_target: Ty<'tcx>,
    root_target: Ty<'tcx>
}
```